homeassistant/components/onboarding/views.py

```diff
@@ -0,0 +1,107 @@
+"""Onboarding views."""
+import asyncio
+
+import voluptuous as vol
+
+from homeassistant.core import callback
+from homeassistant.components.http.view import HomeAssistantView
+from homeassistant.components.http.data_validator import RequestDataValidator
+
+from .const import DOMAIN, STEPS, STEP_USER
+
+
+async def async_setup(hass, data, store):
+    """Setup onboarding."""
+    hass.http.register_view(OnboardingView(data, store))
+    hass.http.register_view(UserOnboardingView(data, store))
+
+
+class OnboardingView(HomeAssistantView):
+    """Returns the onboarding status."""
+
+    requires_auth = False
+    url = '/api/onboarding'
+    name = 'api:onboarding'
+
+    def __init__(self, data, store):
+        """Initialize the onboarding view."""
+        self._store = store
+        self._data = data
+
+    async def get(self, request):
+        """Return the onboarding status."""
+        return self.json([
+            {
+                'step': key,
+                'done': key in self._data['done'],
+            } for key in STEPS
+        ])
+
+
+class _BaseOnboardingView(HomeAssistantView):
+    """Base class for onboarding."""
+
+    requires_auth = False
+    step = None
+
+    def __init__(self, data, store):
+        """Initialize the onboarding view."""
+        self._store = store
+        self._data = data
+        self._lock = asyncio.Lock()
+
+    @callback
+    def _async_is_done(self):
+        """Return if this step is done."""
+        print(self.step, self._data)
+        return self.step in self._data['done']
+
+    async def _async_mark_done(self, hass):
+        """Mark step as done."""
+        self._data['done'].append(self.step)
+        await self._store.async_save(self._data)
+
+        hass.data[DOMAIN] = len(self._data) == len(STEPS)
+
+
+class UserOnboardingView(_BaseOnboardingView):
+    """View to handle onboarding."""
+
+    url = '/api/onboarding/users'
+    name = 'api:onboarding:users'
+    step = STEP_USER
+
+    @RequestDataValidator(vol.Schema({
+        vol.Required('name'): str,
+        vol.Required('username'): str,
+        vol.Required('password'): str,
+    }))
+    async def post(self, request, data):
+        """Return the manifest.json."""
+        hass = request.app['hass']
+
+        async with self._lock:
+            if self._async_is_done():
+                return self.json_message('User step already done', 403)
+
+            provider = _async_get_hass_provider(hass)
```

Comment:
Maybe in the future we can allow the first user coming from other provider by adding setup_shema to the provider.

Reply:
Then we need to check if owner already created in `async_is_onboarded()`

Reply:
It's not a use case I am currently worried about.

Reply:
That's a good point. If an owner exists I'll mark the user step as done.